### DISCUSSION TITLE
ws+tls+cdn搭配优选IP的domain第一次连结会等30秒以上

### DISCUSSION THREAD

**Discussion by kousyougi** [XTLS/Xray-core#674] [Q&A]
我有每天定时更新Cloudflare的优选IP到自己的域名：bestip.domain.com
真实要连的域名是：real.domain.com

客户端设定：

{
　"protocol": "vless",
　"settings": {
　　"vnext": [
　　　{
　　　　"address": "bestip.domain.com",
　　　　"port": 443,
　　　　"users": [{"id": "uuid","encryption": "none"}]
　　　}
　　]
　},
　"streamSettings": {
　　"network": "ws",
　　"security": "tls",
　　"tlsSettings": {
　　　"serverName": "real.domain.com"
　　},
　　"wsSettings": {
　　　"path": "/path?ed=2048",
　　　"headers": {
　　　　"Host": "real.domain.com"
　　　}
　　}
　}
}

vnext的address和 tlsSettings的serverName、headers的Host不同。
这样设定第一次连结，都要耗时30秒以上，网页才会打开。

但如果把优选IP写进/etc/hosts里，或是不用优选IP。
设定檔改成这样：
{
　"protocol": "vless",
　"settings": {
　　"vnext": [
　　　{
　　　　"address": "real.domain.com",
　　　　"port": 443,
　　　　"users": [{"id": "uuid","encryption": "none"}]
　　　}
　　]
　},
　"streamSettings": {
　　"network": "ws",
　　"security": "tls",
　　"wsSettings": {
　　　"path": "/path?ed=2048"
　　}
　}
}

第一次连结网页大概五秒内就能刷出来。

把最新版的v2ray拿来测试也是有一样的问题。

**Comment by kindusmithkindu** [comment]
首次连接需要TLS握手,需要一些时间,这根本没什么关系的

**Comment by kousyougi** [comment]
但两种设定的握手时间差太多。一个5秒以内，一个30秒以上。
复现这个bug很容易，只要把两个domain写进/etc/hosts就能复现。

**Comment by kindusmithkindu** [comment]
你用官方的配置文件试试吧https://github.com/XTLS/Xray-examples/blob/main/VLESS-TCP-TLS-WS%20(recommended)/config_client_ws_tls.json

**Comment by kindusmithkindu** [comment]
serverName和wssetting写了两次域名,等于连一次再连一次,我猜测可能是这个原因

**Comment by kindusmithkindu** [comment]
按照官方的配置,可能是第一次先试着用socks5建立连接,然后目标服务器确认,发回一个tls加密的内容,到本地解密,然后本地通过ws+tls再建立加密连接,连接到目标服务器,你的明显多写了一个,成了3次连接了

**Comment by kousyougi** [comment]
> 按照官方的配置,可能是第一次先试着用socks5建立连接,然后目标服务器确认,发回一个tls加密的内容,到本地解密,然后本地通过ws+tls再建立加密连接,连接到目标服务器,你的明显多写了一个,成了3次连接了

这我都测试过了，当vnext的address跟tls的servername不是相同域名时，如果不写"headers": {"Host": "real.domain.com"}会无法连线。

我会发现这个问题是我同样的设置（vnext address: bestip.domain, tls server name:real.domain, host:real.domain）在ios的小火箭上，竟然第一次连结在五秒以内，才回去排查xray的客户端设置。

**Comment by BozhanL** [comment]
> > 按照官方的配置,可能是第一次先试着用socks5建立连接,然后目标服务器确认,发回一个tls加密的内容,到本地解密,然后本地通过ws+tls再建立加密连接,连接到目标服务器,你的明显多写了一个,成了3次连接了
> 
> 这我都测试过了，当vnext的address跟tls的servername不是相同域名时，如果不写"headers": {"Host": "real.domain.com"}会无法连线。
> 
> 我会发现这个问题是我同样的设置（vnext address: bestip.domain, tls server name:real.domain, host:real.domain）在ios的小火箭上，竟然第一次连结在五秒以内，才回去排查xray的客户端设置。

建议查查DNS解析

**Comment by kousyougi** [comment]
> 你用官方的配置文件试试吧https://github.com/XTLS/Xray-examples/blob/main/VLESS-TCP-TLS-WS%20(recommended)/config_client_ws_tls.json

使用官方文件：当vnext address的domain与tls server name不同时，无法连线。加上host之后就能连线了。但第一次连线时间会超过30秒，就是我想要回报的bug。

**Comment by kousyougi** [comment]
> > > 按照官方的配置,可能是第一次先试着用socks5建立连接,然后目标服务器确认,发回一个tls加密的内容,到本地解密,然后本地通过ws+tls再建立加密连接,连接到目标服务器,你的明显多写了一个,成了3次连接了
> > 
> > 
> > 这我都测试过了，当vnext的address跟tls的servername不是相同域名时，如果不写"headers": {"Host": "real.domain.com"}会无法连线。
> > 我会发现这个问题是我同样的设置（vnext address: bestip.domain, tls server name:real.domain, host:real.domain）在ios的小火箭上，竟然第一次连结在五秒以内，才回去排查xray的客户端设置。
> 
> 建议查查DNS解析

已经写进/etc/hosts了，解析时间相同。
还没写进/etc/hosts前，两个域名用的都是Cloudflare的解析，解析时间都低于一秒。

**Comment by BozhanL** [comment]
> > 按照官方的配置,可能是第一次先试着用socks5建立连接,然后目标服务器确认,发回一个tls加密的内容,到本地解密,然后本地通过ws+tls再建立加密连接,连接到目标服务器,你的明显多写了一个,成了3次连接了
> 
> 这我都测试过了，当vnext的address跟tls的servername不是相同域名时，如果不写"headers": {"Host": "real.domain.com"}会无法连线。
> 
> 我会发现这个问题是我同样的设置（vnext address: bestip.domain, tls server name:real.domain, host:real.domain）在ios的小火箭上，竟然第一次连结在五秒以内，才回去排查xray的客户端设置。

对比两边的配置，看看有什么不同
